src/looprestoration.rs

```diff
@@ -23,3 +30,327 @@ pub union LooprestorationParams {
     pub filter: Align16<[[int16_t; 8]; 2]>,
     pub sgr: LooprestorationParams_sgr,
 }
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct Dav1dLoopRestorationDSPContext {
+    pub wiener: FnVersion,
+    pub sgr: FnVersion,
+}
+
+impl Dav1dLoopRestorationDSPContext {
+    pub unsafe fn call<BD: BitDepthFnLoopRestoration>(
+        &self,
+        filter: FilterFn,
+        mut p: *mut BD::Pixel,
+        stride: ptrdiff_t,
+        left: *const [BD::Pixel; 4],
+        mut lpf: *const BD::Pixel,
+        w: libc::c_int,
+        h: libc::c_int,
+        params: *const LooprestorationParams,
+        edges: LrEdgeFlags,
+        bd: BD,
+    ) {
+        use FilterFn::*;
+
+        let fn_version = match filter {
+            Wiener5 | Wiener7 => self.wiener,
+            Sgr5x5 | Sgr3x3 | SgrMix => self.sgr,
+        };
+
+        bd.call(
+            fn_version, filter, p, stride, left, lpf, w, h, params, edges,
+        )
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum FilterFn {
+    Wiener7,
+    Wiener5,
+    Sgr5x5,
+    Sgr3x3,
+    SgrMix,
+}
+
+impl FilterFn {
+    /// Converts a raw Wiener filter index into the corresponding filter type.
+    pub fn wiener(raw: usize) -> Self {
+        match raw {
+            0 => Self::Wiener7,
+            1 => Self::Wiener5,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Converts a raw SGR filter index into the corresponding filter type.
+    pub fn sgr(raw: usize) -> Self {
+        match raw {
+            0 => Self::Sgr5x5,
+            1 => Self::Sgr3x3,
+            2 => Self::SgrMix,
+            _ => unreachable!(),
+        }
+    }
+}
+
+pub trait BitDepthFnLoopRestoration: BitDepth {
+    unsafe fn call(
+        &self,
```

Comment:
We don't want this publically callable, but `BitDepthFnLoopRestoration` has to be `pub` since it appears in necessarily `pub methods.  In #322, I used a `pub struct Token(())` type with a private constructor, though there might be a cleaner way to accomplish it.